AX: Use ElementName instead of TagName across live and isolated objects

hoffmanjoshua opened the pull request:
#### 1d898339cf40cb4413bb9ace932b45068d399595
<pre>
AX: Use ElementName instead of TagName across live and isolated objects
<a href="https://bugs.webkit.org/show_bug.cgi?id=292996">https://bugs.webkit.org/show_bug.cgi?id=292996</a>
<a href="https://rdar.apple.com/151312403">rdar://151312403</a>

Reviewed by Tyler Wilcock.

Our use of TagName on the isolated tree differed from our live tree tags, which
were qualified names/atom strings. This difference makes moving implementations
to AXCoreObject less graceful.

Moving to ElementName, a uint16_t, for both the live and isolated objects, allows us
to share methods like hasElementName and elementName, without conversions betweeen
types.

Because of this change, many places had to be updated to use Elements, rather than
Nodes, which requires a downcast. This doesn&apos;t add complexity, however, because
previously when using `hasTagName`, we would have to downcast to HTMLElement every
time. There were some cases I optimized as well, to only require one downcast
instead of one per tag-name check, which we would do in the past.

* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::elementName):
(WebCore::isRowGroup):
(WebCore::isAccessibilityList):
(WebCore::AXObjectCache::handleChildrenChanged):
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::traverseToOffsetInRange):
(WebCore::isReplacedNodeOrBR):
(WebCore::characterOffsetNodeIsBR):
(WebCore::canHaveRelations):
(WebCore::AXObjectCache::addLabelForRelation):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::boundaryPoint const):
* Source/WebCore/accessibility/AccessibilityARIAGridRow.cpp:
(WebCore::AccessibilityARIAGridRow::parentTable const):
* Source/WebCore/accessibility/AccessibilityList.cpp:
(WebCore::AccessibilityList::isUnorderedList const):
(WebCore::AccessibilityList::isOrderedList const):
(WebCore::AccessibilityList::isDescriptionList const):
(WebCore::AccessibilityList::determineAccessibilityRoleWithCleanChildren):
* Source/WebCore/accessibility/AccessibilityMathMLElement.cpp:
(WebCore::AccessibilityMathMLElement::determineAccessibilityRole):
(WebCore::AccessibilityMathMLElement::addChildren):
(WebCore::AccessibilityMathMLElement::isMathText const):
(WebCore::AccessibilityMathMLElement::isMathNumber const):
(WebCore::AccessibilityMathMLElement::isMathIdentifier const):
(WebCore::AccessibilityMathMLElement::isMathMultiscript const):
(WebCore::AccessibilityMathMLElement::isMathTable const):
(WebCore::AccessibilityMathMLElement::isMathTableRow const):
(WebCore::AccessibilityMathMLElement::isMathTableCell const):
(WebCore::AccessibilityMathMLElement::mathUnderObject):
(WebCore::AccessibilityMathMLElement::mathOverObject):
(WebCore::AccessibilityMathMLElement::mathSubscriptObject):
(WebCore::AccessibilityMathMLElement::mathSuperscriptObject):
(WebCore::AccessibilityMathMLElement::mathPrescripts):
(WebCore::AccessibilityMathMLElement::mathPostscripts):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):
(WebCore::AccessibilityNodeObject::addChildren):
(WebCore::AccessibilityNodeObject::canHaveChildren const):
(WebCore::AccessibilityNodeObject::computeIsIgnored const):
(WebCore::AccessibilityNodeObject::isNativeImage const):
(WebCore::AccessibilityNodeObject::headingTagLevel const):
(WebCore::AccessibilityNodeObject::isFieldset const):
(WebCore::nodeActionElement):
(WebCore::AccessibilityNodeObject::isGenericFocusableElement const):
(WebCore::AccessibilityNodeObject::captionForFigure const):
(WebCore::AccessibilityNodeObject::usesAltTagForTextComputation const):
(WebCore::AccessibilityNodeObject::visibleText const):
(WebCore::AccessibilityNodeObject::alternativeTextForWebArea const):
(WebCore::AccessibilityNodeObject::title const):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::hasRowGroupTag const):
(WebCore::AccessibilityObject::dependsOnTextUnderElement const):
(WebCore::AccessibilityObject::hasElementName const):
(WebCore::AccessibilityObject::supportsDatetimeAttribute const):
(WebCore::AccessibilityObject::elementName const):
(WebCore::AccessibilityObject::isStyleFormatGroup const):
(WebCore::AccessibilityObject::isFigureElement const):
(WebCore::AccessibilityObject::isOutput const):
(WebCore::AccessibilityObject::hasTagName const): Deleted.
(WebCore::AccessibilityObject::tagName const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::computeIsIgnored const):
(WebCore::AccessibilityRenderObject::rootEditableElementForPosition const):
(WebCore::AccessibilityRenderObject::expandedTextValue const):
(WebCore::AccessibilityRenderObject::supportsExpandedTextValue const):
(WebCore::AccessibilityRenderObject::addCanvasChildren):
* Source/WebCore/accessibility/AccessibilitySVGObject.cpp:
(WebCore::AccessibilitySVGObject::inheritsPresentationalRole const):
* Source/WebCore/accessibility/AccessibilityTable.cpp:
(WebCore::AccessibilityTable::isDataTable const):
(WebCore::AccessibilityTable::computeCellSlots):
* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
(WebCore::AccessibilityTableCell::computeIsIgnored const):
(WebCore::AccessibilityTableCell::isTableHeaderCell const):
(WebCore::AccessibilityTableCell::isColumnHeader const):
(WebCore::AccessibilityTableCell::isRowHeader const):
(WebCore::AccessibilityTableCell::titleUIElement const):
* Source/WebCore/accessibility/atspi/AccessibilityObjectAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::attributes const):
* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
(WebCore::blockquoteLevel):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper isSVGGroupElement]):
(-[WebAccessibilityObjectWrapper detailParentForSummaryObject:]):
(-[WebAccessibilityObjectWrapper detailParentForObject:]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setProperty):
(WebCore::AXIsolatedObject::hasRowGroupTag const):
(WebCore::AXIsolatedObject::headingTagLevel const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::AccessibilityObject::subrolePlatformString const):

Canonical link: <a href="https://commits.webkit.org/295005@main">https://commits.webkit.org/295005@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b715ef12b4f53904958430a4a8b6c8678826e672

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103768 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23470 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13790 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108959 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54420 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105808 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23819 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32016 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78834 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106774 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18483 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93608 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59168 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18295 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11657 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53796 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88068 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11714 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111348 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30924 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22800 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87837 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31285 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89809 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87490 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22279 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32381 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10099 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25281 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30852 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36155 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30646 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33981 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32207 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -3341,7 +3341,7 @@ CharacterOffset AXObjectCache::traverseToOffsetInRange(const SimpleRange& range,
                             continue;
                         }
-                    } else if (previousNode && previousNode->isTextNode() && previousNode->isDescendantOf(currentNode) && currentNode->hasTagName(pTag)) {
+                    } else if (previousNode && previousNode->isTextNode() && previousNode->isDescendantOf(currentNode) && is<Element>(currentNode) && uncheckedDowncast<Element>(currentNode)->elementName() == ElementName::HTML_p) {
```
Could we add a new `Node::hasElementName` method for cases like these? It would avoid adding a ton more length to this line and others.

hoffmanjoshua replied:
Tyler and I discussed offline to add a global function, `WebCore::elementName(Node*)` to get the element name more easily.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.h:
```diff
@@ -39,6 +39,12 @@ class Element;
 class HTMLLabelElement;
 class Node;
 
+struct ElementNameHash {
```
What requires the addition of this struct?

hoffmanjoshua replied:
Sorry, this shouldn't be here anymore!

twilco commented on Source/WebCore/accessibility/AccessibilityTableCell.cpp:
```diff
@@ -152,20 +152,20 @@ AccessibilityRole AccessibilityTableCell::determineAccessibilityRole()
         return false;
 
-    if (node->hasTagName(thTag))
+    if (element->elementName() == ElementName::HTML_th)
```
Can we get the elementName() once instead of doing it multiple times? That would be more succinct, and more performant.

twilco commented on Source/WebCore/accessibility/AccessibilityTableCell.cpp:
```diff
@@ -216,15 +216,15 @@ bool AccessibilityTableCell::isRowHeader() const
-    for (RefPtr ancestor = node()->parentNode(); ancestor; ancestor = ancestor->parentNode()) {
-        if (ancestor->hasTagName(tfootTag) || ancestor->hasTagName(tbodyTag) || ancestor->hasTagName(tableTag)) {
+    for (RefPtr ancestor = element()->parentElement(); ancestor; ancestor = ancestor->parentElement()) {
+        if (ancestor->elementName() == ElementName::HTML_tfoot || ancestor->elementName() == ElementName::HTML_tbody || ancestor->elementName() == ElementName::HTML_table) {
```
Can we get elementName once?

twilco commented on Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
```diff
@@ -245,8 +245,8 @@ static unsigned blockquoteLevel(RenderObject* renderer)
     unsigned result = 0;
-    for (Node* node = renderer->node(); node; node = node->parentNode()) {
-        if (node->hasTagName(HTMLNames::blockquoteTag))
+    for (Element* element = dynamicDowncast<Element>(renderer->node()); element; element = element->parentElement()) {
```
Changing to `parentElement()` iteration rather than `parentNode` iteration has subtly different behavior — if the parent is a node, not an element, we'll stop iterating, even if there are more ancestors above it. One example where this would be relevant is for `DocumentFragment`s, which are `ContainerNode`s, and thus can have children, but are not elements.

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -5405,7 +5411,7 @@ bool AXObjectCache::addRelation(Element& origin, Element& target, AXRelation rel
 static bool canHaveRelations(Element& element)
 {
-    return !(element.hasTagName(metaTag) || element.hasTagName(headTag) || element.hasTagName(scriptTag) || element.hasTagName(htmlTag) || element.hasTagName(styleTag));
+    return !(element.elementName() == ElementName::HTML_meta || element.elementName() == ElementName::HTML_head || element.elementName() == ElementName::HTML_script || element.elementName() == ElementName::HTML_html || element.elementName() == ElementName::HTML_style);
```
Can we get element name once to avoid constantly re-reading it?

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -3482,7 +3488,7 @@ std::optional<SimpleRange> AXObjectCache::rangeMatchesTextNearRange(const Simple
 static bool isReplacedNodeOrBR(Node& node)
 {
-    return AccessibilityObject::replacedNodeNeedsCharacter(node) || node.hasTagName(brTag);
+    return AccessibilityObject::replacedNodeNeedsCharacter(node) || WebCore::elementName(&node) == ElementName::HTML_br;
```
Maybe we could have an `elementName` function that takes a `Node&` too, since this is now going to incur an extra null-check in the cast to element. We use this pattern elsewhere, e.g. for `hasAnyRole`:

https://github.com/WebKit/WebKit/blob/798647db3b710d3f5282bbc367a44bcc44200173/Source/WebCore/accessibility/AXObjectCache.cpp#L663#L666

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -194,7 +194,7 @@ std::optional<BoundaryPoint> AXTextMarker::boundaryPoint() const
 
     int offset = characterOffset.startIndex + characterOffset.offset;
-    if (AccessibilityObject::replacedNodeNeedsCharacter(*node) || node->hasTagName(HTMLNames::brTag))
+    if (AccessibilityObject::replacedNodeNeedsCharacter(*node) || WebCore::elementName(node.get()) == ElementName::HTML_br)
```
Could use the `Node&` version of `elementName` here (the prior code dereferenced it unconditionally)

twilco commented on Source/WebCore/accessibility/AccessibilityList.cpp:
```diff
@@ -178,7 +176,7 @@ AccessibilityRole AccessibilityList::determineAccessibilityRoleWithCleanChildren
                     hasVisibleMarkers = true;
                 listItemCount++;
-            } else if (node && node->hasTagName(liTag)) {
+            } else if (node && WebCore::elementName(node.get()) == ElementName::HTML_li) {
```
Could use the Node& version of elementName here to avoid duplicating the null-check

twilco commented on Source/WebCore/accessibility/AccessibilityMathMLElement.cpp:
```diff
@@ -59,7 +59,7 @@ AccessibilityRole AccessibilityMathMLElement::determineAccessibilityRole()
 
     Node* node = m_renderer->node();
-    if (node && node->hasTagName(MathMLNames::mathTag))
+    if (node && WebCore::elementName(node) == ElementName::MathML_math)
```
Could use the Node& version of elementName here to avoid duplicating the null-check

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -351,66 +352,66 @@ AccessibilityRole AccessibilityNodeObject::determineAccessibilityRoleFromNode(Tr
+    if (elementName == ElementName::HTML_code)
         return AccessibilityRole::Code;
-    if (element->hasTagName(delTag) || element->hasTagName(sTag))
+    if (elementName == ElementName::HTML_del || element->elementName() == ElementName::HTML_s)
```
`element->elementName() == ElementName::HTML_s`

Don't we already have an `elementName` variable? Why re-get it?

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -607,7 +608,7 @@ void AccessibilityNodeObject::addChildren()
 
     // The only time we add children from the DOM tree to a node with a renderer is when it's a canvas.
-    if (renderer() && !node->hasTagName(canvasTag))
+    if (renderer() && WebCore::elementName(node.get()) != ElementName::HTML_canvas)
```
Could use the `Node&` version of `elementName` (the previous code dereferenced it unconditionally)

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -703,7 +704,7 @@ bool AccessibilityNodeObject::computeIsIgnored() const
         RefPtr parent = node->parentNode();
         // Fallback content in iframe nodes should be ignored.
-        if (parent && parent->hasTagName(iframeTag) && parent->renderer())
+        if (parent && WebCore::elementName(parent.get()) == ElementName::HTML_iframe && parent->renderer())
```
Could use the Node& version of `elementName` to avoid duplicating the null-check.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -985,18 +987,18 @@ AXCoreObject::AccessibilityChildrenVector AccessibilityNodeObject::radioButtonGr
 {
-    const auto& tag = tagName();
-    if (tag == h1Tag)
+    const auto& elementName = this->elementName();
```
This probably shouldn't be `const auto&` anymore since the return type is uint16_t rather than an `const AtomString&` like tagName was.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1090,7 +1092,7 @@ bool AccessibilityNodeObject::isFieldset() const
         return false;
 
-    return node->hasTagName(fieldsetTag);
+    return WebCore::elementName(node) == ElementName::HTML_fieldset;
```
Can shorten this whole function:

```cpp
bool AccessibilityNodeObject::isFieldset() const
{
    return elementName() == ElementName::HTML_fieldset;
}
```

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1568,7 +1571,7 @@ bool AccessibilityNodeObject::isGenericFocusableElement() const
     if (role == AccessibilityRole::WebArea)
         return false;
-    if (node() && node()->hasTagName(bodyTag))
+    if (node() && WebCore::elementName(node()) == ElementName::HTML_body)
```
Can shorten this to:

`if (elementName() == ElementName::HTML_body)`

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1609,15 +1612,15 @@ AccessibilityObject* AccessibilityNodeObject::captionForFigure() const
     Node* node = this->node();
     for (Node* child = node->firstChild(); child; child = child->nextSibling()) {
-        if (child->hasTagName(figcaptionTag))
+        if (WebCore::elementName(child) == ElementName::HTML_figcaption)
```
Can use the `Node&` version of `elementName` to avoid the duplicated null-check

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1864,7 +1867,7 @@ void AccessibilityNodeObject::visibleText(Vector<AccessibilityText>& textOrder)
 
     // If this node isn't rendered, there's no inner text we can extract from a select element.
-    if (!isAccessibilityRenderObject() && node->hasTagName(selectTag))
+    if (!isAccessibilityRenderObject() && WebCore::elementName(node.get()) == ElementName::HTML_select)
```
Can use the Node& version of elementName to avoid the duplicated null-check

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1963,7 +1966,7 @@ String AccessibilityNodeObject::alternativeTextForWebArea() const
 
     if (auto* owner = document->ownerElement()) {
-        if (owner->hasTagName(frameTag) || owner->hasTagName(iframeTag)) {
+        if (owner->elementName() == ElementName::HTML_frame || owner->elementName() == ElementName::HTML_iframe) {
```
Can we get the elementName once? Would probably be shorter and avoid re-reading the same memory multiple times.

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2324,7 +2327,7 @@ String AccessibilityNodeObject::title() const
 
     // For <select> elements, title should be empty if they are not rendered or have role PopUpButton.
-    if (node && node->hasTagName(selectTag)
+    if (node && WebCore::elementName(node.get()) == ElementName::HTML_select
```
Can use the Node& version of elementName to avoid the duplicated null-check

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -1914,8 +1915,8 @@ std::optional<VisiblePosition> AccessibilityObject::previousLineStartPositionInt
 {
-    const auto& tag = tagName();
-    return tag == theadTag || tag == tbodyTag || tag == tfootTag;
+    const auto& name = elementName();
```
This probably shouldn't be `const auto&` now that we're returning an `ElementName` (`uint16_t`) rather than an `const AtomString&` like `tagName()` used to.

twilco commented on Source/WebCore/accessibility/AccessibilityObject.h:
```diff
@@ -564,13 +564,13 @@ class AccessibilityObject : public AXCoreObject, public CanMakeWeakPtr<Accessibi
-    bool hasAttachmentTag() const final { return hasTagName(HTMLNames::attachmentTag); }
-    bool hasBodyTag() const final { return hasTagName(HTMLNames::bodyTag); }
-    bool hasMarkTag() const final { return hasTagName(HTMLNames::markTag); }
+    bool hasElementName(const ElementName) const override;
```
Can this be `final`? Or does some other subclass override it?

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1345,11 +1345,11 @@ bool AccessibilityRenderObject::computeIsIgnored() const
     // the side effect of causing the immediate parent accessible to be ignored. This is especially
     // problematic for platforms which have distinct roles for textual block elements.
-    if (node && node->hasTagName(spanTag))
+    if (node && WebCore::elementName(node) == ElementName::HTML_span)
```
Can use the Node& version of elementName to avoid the duplicated null-check

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -1345,11 +1345,11 @@ bool AccessibilityRenderObject::computeIsIgnored() const
 
     // Other non-ignored host language elements
-    if (node && node->hasTagName(dfnTag))
+    if (node && WebCore::elementName(node) == ElementName::HTML_dfn)
```
Can use the Node& version of elementName to avoid the duplicated null-check

twilco commented on Source/WebCore/accessibility/AccessibilityTable.cpp:
```diff
@@ -167,13 +167,13 @@ bool AccessibilityTable::isDataTable() const
         // If there's a colgroup or col element, it's probably a data table.
         for (const auto& child : childrenOfType<HTMLElement>(*tableElement)) {
-            if (child.hasTagName(colTag) || child.hasTagName(colgroupTag))
+            if (child.elementName() == ElementName::HTML_col || child.elementName() == ElementName::HTML_colgroup)
```
Can we get the elementName once? Would probably be shorter and avoid re-reading the same memory multiple times.

twilco commented on Source/WebCore/accessibility/AccessibilityTable.cpp:
```diff
@@ -239,13 +239,13 @@ bool AccessibilityTable::isDataTable() const
         for (RefPtr currentElement = currentParent ? currentParent->firstElementChild() : nullptr; currentElement; currentElement = currentElement->nextElementSibling()) {
             if (auto* tableSectionElement = dynamicDowncast<HTMLTableSectionElement>(currentElement.get())) {
-                if (tableSectionElement->hasTagName(theadTag)) {
+                if (tableSectionElement->elementName() == ElementName::HTML_thead) {
```
Can we get the elementName for tableSectionElement  once? Would probably be shorter and avoid re-reading the same memory multiple times.

twilco commented on Source/WebCore/accessibility/AccessibilityTableCell.cpp:
```diff
@@ -156,14 +156,15 @@ bool AccessibilityTableCell::isTableHeaderCell() const
         // i < 2 is used here because in a properly structured table, the thead should be 2 levels away from the td.
         for (int i = 0; i < 2 && current; i++) {
-            if (current->hasTagName(theadTag))
+            if (WebCore::elementName(current) == ElementName::HTML_thead)
```
Can use the Node& version of elementName to avoid the duplicated null-check

twilco commented on Source/WebCore/accessibility/AccessibilityTableCell.cpp:
```diff
@@ -187,11 +188,12 @@ bool AccessibilityTableCell::isColumnHeader() const
     // Checking tableTag and tbodyTag allows to check the case of direct row placement in the table and lets stop the loop at the table level.
     for (RefPtr ancestor = node()->parentNode(); ancestor; ancestor = ancestor->parentNode()) {
-        if (ancestor->hasTagName(theadTag))
+        auto elementName = WebCore::elementName(ancestor.get());
```
Can use the Node& version of elementName to avoid the duplicated null-check

twilco commented on Source/WebCore/accessibility/AccessibilityTableCell.cpp:
```diff
@@ -217,14 +219,15 @@ bool AccessibilityTableCell::isRowHeader() const
     // Checking tableTag allows to check the case of direct row placement in the table and lets stop the loop at the table level.
     for (RefPtr ancestor = node()->parentNode(); ancestor; ancestor = ancestor->parentNode()) {
-        if (ancestor->hasTagName(tfootTag) || ancestor->hasTagName(tbodyTag) || ancestor->hasTagName(tableTag)) {
+        auto elementName = WebCore::elementName(ancestor.get());
```
Can use the Node& version of elementName to avoid the duplicated null-check

twilco commented on Source/WebCore/accessibility/AccessibilityTableCell.cpp:
```diff
@@ -322,7 +325,7 @@ AccessibilityObject* AccessibilityTableCell::titleUIElement() const
     // they are title ui elements
     Node* node = m_renderer->node();
-    if (node && node->hasTagName(thTag))
+    if (node && WebCore::elementName(node) == ElementName::HTML_th)
```
Can be shortened to:

`if (WebCore::elementName(node) == ElementName::HTML_th)`

twilco replied:
Can be shortened to:

`if (WebCore::elementName(node) == ElementName::HTML_th)`

twilco commented on Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
```diff
@@ -246,7 +246,7 @@ static unsigned blockquoteLevel(RenderObject* renderer)
     unsigned result = 0;
     for (Node* node = renderer->node(); node; node = node->parentNode()) {
-        if (node->hasTagName(HTMLNames::blockquoteTag))
+        if (WebCore::elementName(node) == ElementName::HTML_blockquote)
```
Can use the Node& version of elementName to avoid the duplicated null-check

twilco commented on Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
```diff
@@ -870,7 +870,7 @@ - (BOOL)isSVGGroupElement
     // If an SVG group element has a title, it should be an accessible element on iOS.
     Node* node = self.axBackingObject->node();
-    if (node && node->hasTagName(SVGNames::gTag) && [[self accessibilityLabel] length] > 0)
+    if (node && WebCore::elementName(node) == ElementName::SVG_g && [[self accessibilityLabel] length] > 0)
```
Can be shortened to:

`if (WebCore::elementName(node) == ElementName::SVG_g && [[self accessibilityLabel] length] > 0)`

twilco commented on Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
```diff
@@ -2107,19 +2107,19 @@ - (AXCoreObject*)detailParentForSummaryObject:(AccessibilityObject&)object
     return Accessibility::findAncestor<AccessibilityObject>(object, true, [&] (const AccessibilityObject& object) {
-        const auto& tag = object.tagName();
-        if (tag == summaryTag)
+        const auto& elementName = object.elementName();
```
This probably shouldn't be `const auto&` anymore since the return type is uint16_t rather than an `const AtomString&` like tagName was.

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -2337,19 +2339,19 @@ AXIsolatedObject* AXIsolatedObject::headerContainer()
-    TagName tag = propertyValue<TagName>(AXProperty::TagName);
-    switch (tag) {
-    case TagName::h1:
+    ElementName name = propertyValue<ElementName>(AXProperty::ElementName);
```
`auto` seems appropriate here since we specify the type of `ElementName` already as a template parameter to `propertyValue`

Also, any reason we don't just call the `elementName()` function you added?

twilco commented on Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
```diff
@@ -394,16 +394,16 @@
     using namespace HTMLNames;
-    const auto& tag = tagName();
-    if (tag == kbdTag)
+    const auto& elementName = this->elementName();
```
This probably shouldn't be const auto& anymore since the return type is uint16_t rather than an const AtomString& like tagName was.

hoffmanjoshua committed:
AX: Use ElementName instead of TagName across live and isolated objects
https://bugs.webkit.org/show_bug.cgi?id=292996
rdar://151312403

Reviewed by Tyler Wilcock.

Our use of TagName on the isolated tree differed from our live tree tags, which
were qualified names/atom strings. This difference makes moving implementations
to AXCoreObject less graceful.

Moving to ElementName, a uint16_t, for both the live and isolated objects, allows us
to share methods like hasElementName and elementName, without conversions betweeen
types.

Because of this change, many places had to be updated to use Elements, rather than
Nodes, which requires a downcast. This doesn't add complexity, however, because
previously when using `hasTagName`, we would have to downcast to HTMLElement every
time. There were some cases I optimized as well, to only require one downcast
instead of one per tag-name check, which we would do in the past.

* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::elementName):
(WebCore::isRowGroup):
(WebCore::isAccessibilityList):
(WebCore::AXObjectCache::handleChildrenChanged):
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::traverseToOffsetInRange):
(WebCore::isReplacedNodeOrBR):
(WebCore::characterOffsetNodeIsBR):
(WebCore::canHaveRelations):
(WebCore::AXObjectCache::addLabelForRelation):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::boundaryPoint const):
* Source/WebCore/accessibility/AccessibilityARIAGridRow.cpp:
(WebCore::AccessibilityARIAGridRow::parentTable const):
* Source/WebCore/accessibility/AccessibilityList.cpp:
(WebCore::AccessibilityList::isUnorderedList const):
(WebCore::AccessibilityList::isOrderedList const):
(WebCore::AccessibilityList::isDescriptionList const):
(WebCore::AccessibilityList::determineAccessibilityRoleWithCleanChildren):
* Source/WebCore/accessibility/AccessibilityMathMLElement.cpp:
(WebCore::AccessibilityMathMLElement::determineAccessibilityRole):
(WebCore::AccessibilityMathMLElement::addChildren):
(WebCore::AccessibilityMathMLElement::isMathText const):
(WebCore::AccessibilityMathMLElement::isMathNumber const):
(WebCore::AccessibilityMathMLElement::isMathIdentifier const):
(WebCore::AccessibilityMathMLElement::isMathMultiscript const):
(WebCore::AccessibilityMathMLElement::isMathTable const):
(WebCore::AccessibilityMathMLElement::isMathTableRow const):
(WebCore::AccessibilityMathMLElement::isMathTableCell const):
(WebCore::AccessibilityMathMLElement::mathUnderObject):
(WebCore::AccessibilityMathMLElement::mathOverObject):
(WebCore::AccessibilityMathMLElement::mathSubscriptObject):
(WebCore::AccessibilityMathMLElement::mathSuperscriptObject):
(WebCore::AccessibilityMathMLElement::mathPrescripts):
(WebCore::AccessibilityMathMLElement::mathPostscripts):
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):
(WebCore::AccessibilityNodeObject::addChildren):
(WebCore::AccessibilityNodeObject::canHaveChildren const):
(WebCore::AccessibilityNodeObject::computeIsIgnored const):
(WebCore::AccessibilityNodeObject::isNativeImage const):
(WebCore::AccessibilityNodeObject::headingTagLevel const):
(WebCore::AccessibilityNodeObject::isFieldset const):
(WebCore::nodeActionElement):
(WebCore::AccessibilityNodeObject::isGenericFocusableElement const):
(WebCore::AccessibilityNodeObject::captionForFigure const):
(WebCore::AccessibilityNodeObject::usesAltTagForTextComputation const):
(WebCore::AccessibilityNodeObject::visibleText const):
(WebCore::AccessibilityNodeObject::alternativeTextForWebArea const):
(WebCore::AccessibilityNodeObject::title const):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::hasRowGroupTag const):
(WebCore::AccessibilityObject::dependsOnTextUnderElement const):
(WebCore::AccessibilityObject::hasElementName const):
(WebCore::AccessibilityObject::supportsDatetimeAttribute const):
(WebCore::AccessibilityObject::elementName const):
(WebCore::AccessibilityObject::isStyleFormatGroup const):
(WebCore::AccessibilityObject::isFigureElement const):
(WebCore::AccessibilityObject::isOutput const):
(WebCore::AccessibilityObject::hasTagName const): Deleted.
(WebCore::AccessibilityObject::tagName const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::computeIsIgnored const):
(WebCore::AccessibilityRenderObject::rootEditableElementForPosition const):
(WebCore::AccessibilityRenderObject::expandedTextValue const):
(WebCore::AccessibilityRenderObject::supportsExpandedTextValue const):
(WebCore::AccessibilityRenderObject::addCanvasChildren):
* Source/WebCore/accessibility/AccessibilitySVGObject.cpp:
(WebCore::AccessibilitySVGObject::inheritsPresentationalRole const):
* Source/WebCore/accessibility/AccessibilityTable.cpp:
(WebCore::AccessibilityTable::isDataTable const):
(WebCore::AccessibilityTable::computeCellSlots):
* Source/WebCore/accessibility/AccessibilityTableCell.cpp:
(WebCore::AccessibilityTableCell::computeIsIgnored const):
(WebCore::AccessibilityTableCell::isTableHeaderCell const):
(WebCore::AccessibilityTableCell::isColumnHeader const):
(WebCore::AccessibilityTableCell::isRowHeader const):
(WebCore::AccessibilityTableCell::titleUIElement const):
* Source/WebCore/accessibility/atspi/AccessibilityObjectAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::attributes const):
* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
(WebCore::blockquoteLevel):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper isSVGGroupElement]):
(-[WebAccessibilityObjectWrapper detailParentForSummaryObject:]):
(-[WebAccessibilityObjectWrapper detailParentForObject:]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setProperty):
(WebCore::AXIsolatedObject::hasRowGroupTag const):
(WebCore::AXIsolatedObject::headingTagLevel const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::AccessibilityObject::subrolePlatformString const):

Canonical link: https://commits.webkit.org/295005@main
